### PR TITLE
Filter out topics that aren't available in the page's language

### PR DESCRIPTION
**Pull request by franklinsch**
Bug/issue #, if applicable: rdar://89124291

## Summary

Filters out topics that aren't available in the language of the symbol that curates them. Note that this was already working for automatically curated topics, but not manually curated topics.

This PR also includes making tutorial content available on all the languages in which the framework is available (last commit),  

Also, since the original fix of this PR filters out topics that aren't available in the parent page's language, it was also causing DocC to filter out tutorial content from Objective-C pages. As such, the last commit of this PR resolves this issue by making tutorials multi-language as well, if the framework it's documenting is multi-language.

## Dependencies

None.

## Testing

1. Create symbol graph files with symbols that are only available in Swift and symbols only available in Objective-C, and curate them manually in a documentation extension file or in documentation comments.
2. Verify that the rendered documentation doesn't show the Objective-C–only symbols when browsing the Swift documentation, and vice-versa.

## Checklist

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [x] Updated documentation if necessary

## TODO

- [x] Make Tutorials multi-language so that they can be curated in Objective-C pages. They're currently hardcoded to be Swift-only

### CODE REVIEW

**Comment by franklinsch** [comment]
@swift-ci please test

**Comment by franklinsch** [comment]
@swift-ci please test

**Comment by franklinsch** [comment]
@swift-ci please test

File: Sources/SwiftDocC/Utility/DataStructures/GroupedSequence.swift [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    
+    /// A closure that transforms an element into its key.
+    private var deriveKey: (Element) -> Key
+    
+    var description: String {
+        storage.values.description
+    }
+    
+    /// Creates an empty grouped sequence.
+    init(deriveKey: @escaping (Element) -> Key) {
```
Comment (daniel-grumberg) [inline Sources/SwiftDocC/Utility/DataStructures/GroupedSequence.swift:52]: Nit: typo `derivedKey`

File: Sources/SwiftDocC/Utility/DataStructures/GroupedSequence.swift [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+/// ```swift
+/// for item in groupedSequence {
+///     print(item)
+/// }
+///
+/// // Prints "aa"
+/// // Prints "a"
+/// // Prints "aaa"
+/// ```
+struct GroupedSequence<Key: Hashable, Element>: Sequence, CustomStringConvertible {
```
Comment (daniel-grumberg) [inline Sources/SwiftDocC/Utility/DataStructures/GroupedSequence.swift:41]: Not needed for this PR but this already fulfills the requirements for `Collection` might be nice since we are making this a standalone data structure.

File: Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift [hunk 3]
```diff
@@ -823,6 +850,20 @@ public struct RenderNodeTranslator: SemanticVisitor {
                 return ContentRenderSection(kind: .content, content: discussionContent, heading: "Discussion")
             }
             
+            /// Returns whether the topic with the given identifier is available in one of the traits in `allowedTraits`.
+            func isAvailableInAllowedTrait(identifier: String) -> Bool {
+                guard let reference = contentCompiler.collectedTopicReferences[identifier] else {
```
Comment (daniel-grumberg) [inline Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift:855]: Not sure I understand this logic. Could you explain why you return true if you couldn't find a topic reference for this identifier?

File: Tests/SwiftDocCTests/Utility/DataStructures/GroupedSequenceTests.swift [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+class GroupedSequenceTests: XCTestCase {
+    /// A grouped sequence of strings, with their number of characters as the key.
+    var groupedSequence = GroupedSequence<Int, String>(deriveKey: \.count)
+    
+    func testAppends() {
+        groupedSequence.append("a")
+        groupedSequence.append("aa")
+        
+        XCTAssertEqual(groupedSequence[1], "a")
+        XCTAssertEqual(groupedSequence[2], "aa")
```
Comment (daniel-grumberg) [inline Tests/SwiftDocCTests/Utility/DataStructures/GroupedSequenceTests.swift:23]: you probably want to add the following to this:
```
groupedSequence.append("b")
XCTAssertEqual(groupedSequence[1], "b")
```

File: Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift (hunk 3 above)
Comment (daniel-grumberg) [reply Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift:855]: Is this because you are conservatively not filtering out content you don't have a definition for in this framework?

File: Sources/SwiftDocC/Utility/DataStructures/GroupedSequence.swift (hunk 1 above)
Comment (franklinsch) [reply Sources/SwiftDocC/Utility/DataStructures/GroupedSequence.swift:52]: I think `deriveKey` as it is currently works better because its type is a function, so it should be name like an action.

File: Sources/SwiftDocC/Utility/DataStructures/GroupedSequence.swift (hunk 2 above)
Comment (franklinsch) [reply Sources/SwiftDocC/Utility/DataStructures/GroupedSequence.swift:41]: hm, yes, looks like we're close! But we need to implement a few more requirements, which I think we should implement when we get a use for them. This isn't intended to be a general-purpose API—it's just for DocC's purposes.

File: Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift (hunk 3 above)
Comment (franklinsch) [reply Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift:855]: Ah yes, I should clarify that in a comment, thanks for pointing that out. It's expected for `reference` to never be `nil` as it is a precondition for `contentCompiler` to have populated `collectionTopicReference` with the reference at this point. Otherwise, it is a bug in `contentCompiler`.

File: Tests/SwiftDocCTests/Utility/DataStructures/GroupedSequenceTests.swift (hunk 4 above)
Comment (franklinsch) [reply Tests/SwiftDocCTests/Utility/DataStructures/GroupedSequenceTests.swift:23]: Yes, that's a great idea!

File: Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift (hunk 3 above)
Comment (franklinsch) [reply Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift:855]: In fact, we can actually add an [`assertionFailure`](https://developer.apple.com/documentation/swift/1539616-assertionfailure) so that we get a crash in debug builds, but not release.

Comment (franklinsch) [reply Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift:855]: Actually, after taking a closer look, I'm wrong on this. It's possible for the content compiler to generate a reference identifier without registering it in `collectedTopicReferences`, for example when the reference is a non-doc URL (e.g., `https://` in which case it gets registered in `linkReferences` instead. I'll add a comment to clarify that.

**Comment by franklinsch** [comment]
@swift-ci please teset

**Comment by franklinsch** [comment]
@swift-ci please test

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift [hunk 5]
```diff
@@ -621,9 +624,19 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
...
                 
-                let sourceLanguage = SourceLanguage.swift
-                let node = DocumentationNode(reference: technologyResult.topicGraphNode.reference, kind: .technology, sourceLanguage: sourceLanguage, name: .conceptual(title: technology.intro.title), markup: technology.originalMarkup, semantic: technology)
-                documentationCache[technologyResult.topicGraphNode.reference] = node
+                let reference = technologyResult.topicGraphNode.reference.withSourceLanguages(sourceLanguages)
+                
+                let node = DocumentationNode(
+                    reference: reference,
+                    kind: .technology,
+                    sourceLanguage: Self.defaultLanguage(in: sourceLanguages),
```
Comment (ethan-kusters) [inline Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:632]: Nit since I don't think this is directly related to this PR. But this should this be `defaultSourceLanguage`?

```suggestion
                    defaultSourceLanguage: Self.defaultLanguage(in: sourceLanguages),
```

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift [hunk 6]
```diff
@@ -621,9 +624,19 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
...
+                    availableSourceLanguages: sourceLanguages,
+                    name: .conceptual(title: technology.intro.title),
+                    markup: technology.originalMarkup,
+                    semantic: technology
+                )
+                documentationCache[reference] = node
+                topicGraph.nodes[reference]?.reference = reference
```
Comment (ethan-kusters) [inline Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:639]: Is this a change from before? We weren't including technology references in the topic graph?

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift [hunk 7]
```diff
@@ -621,9 +624,19 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
                 // Add to document map
                 documentLocationMap[url] = technologyResult.topicGraphNode.reference
                 
-                let sourceLanguage = SourceLanguage.swift
-                let node = DocumentationNode(reference: technologyResult.topicGraphNode.reference, kind: .technology, sourceLanguage: sourceLanguage, name: .conceptual(title: technology.intro.title), markup: technology.originalMarkup, semantic: technology)
-                documentationCache[technologyResult.topicGraphNode.reference] = node
+                let reference = technologyResult.topicGraphNode.reference.withSourceLanguages(sourceLanguages)
```
Comment (ethan-kusters) [inline Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:627]: Readability nit:
```suggestion
                let technologyReference = technologyResult.topicGraphNode.reference.withSourceLanguages(sourceLanguages)
```

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift [hunk 8]
```diff
@@ -621,9 +624,19 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
...
+                
+                let node = DocumentationNode(
```
Comment (ethan-kusters) [inline Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:629]: Readability nit:
```suggestion
                let technologyNode = DocumentationNode(
```

Probably partly due to the GitHub UI but this diff was generally hard for me to follow because we have a lot of otherwise unspecified `node`s and `reference`.

File: Sources/SwiftDocC/Infrastructure/Topic Graph/TopicGraph.swift [hunk 9]
```diff
@@ -70,7 +70,7 @@ struct TopicGraph {
         }
         
         /// The reference to the `DocumentationNode` this node represents.
-        let reference: ResolvedTopicReference
+        var reference: ResolvedTopicReference
```
Comment (ethan-kusters) [inline Sources/SwiftDocC/Infrastructure/Topic Graph/TopicGraph.swift:73]: It seems strange to me to make only the reference mutable. Should we just be making a new DocumentationNode when we need to modify the reference? That seems like a core part of defines the node.

File: Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift [hunk 10]
```diff
@@ -129,6 +129,9 @@ public struct RenderNodeTranslator: SemanticVisitor {
         node.hierarchy = hierarchy.hierarchy
         node.metadata.category = technology.name
         
+        let documentationNode = try! context.entity(with: identifier)
```
Comment (ethan-kusters) [inline Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift:132]: Definitely out-of-scope for this PR but it seems like the `visitTutorial` method should `throw` at this point.

File: Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift [hunk 11]
```diff
@@ -823,6 +850,22 @@ public struct RenderNodeTranslator: SemanticVisitor {
                 return ContentRenderSection(kind: .content, content: discussionContent, heading: "Discussion")
             }
             
+            /// Returns whether the topic with the given identifier is available in one of the traits in `allowedTraits`.
+            func isAvailableInAllowedTrait(identifier: String) -> Bool {
```
Comment (ethan-kusters) [inline Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift:854]:
```suggestion
            func topicIsAvailableInAllowedTraits(identifier topicIdentifier: String) -> Bool {
```

File: Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift [hunk 12]
```diff
@@ -1011,7 +1068,7 @@ public struct RenderNodeTranslator: SemanticVisitor {
                     DeclarationsSectionTranslator(),
                     ReturnsSectionTranslator(),
                     ParametersSectionTranslator(),
-                    DiscussionSectionTranslator()
+                    DiscussionSectionTranslator(),
```
Comment (ethan-kusters) [inline Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift:1071]: 😃

**Review by ethan-kusters — COMMENTED** [review]
This looks good to me! Just a few small things for your consideration.

Was tutorials the last thing we needed to support here or do we still need support for articles?

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift (hunk 5 above)
Comment (franklinsch) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:632]: Hmm, possibly. Out of scope of this PR though, especially as `sourceLanguage` is public API. Thanks for bringing it up!

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift (hunk 6 above)
Comment (franklinsch) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:639]: We are at an earlier stage, in `registerDocuments`. What we're doing here is updating the reference now that we have enough information to know in which languages the page is available in. I'll add a comment to clarify that.

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift (hunk 8 above)
Comment (franklinsch) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:629]: Makes sense, I see it's hard to follow in the GitHub diff as well. I'll take your suggestions!

File: Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift (hunk 10 above)
Comment (franklinsch) [reply Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift:132]: In this case, it's actually a programmer error if `context.entity(with:)` throws, so the force-unwrap makes sense. It could be improved with a `preconditionFailure` with a more user-facing message, though.

File: Sources/SwiftDocC/Infrastructure/Topic Graph/TopicGraph.swift (hunk 9 above)
Comment (franklinsch) [reply Sources/SwiftDocC/Infrastructure/Topic Graph/TopicGraph.swift]: I think since `Node` is a reference type, it makes sense to avoid accidental mutation. I added an API to create a new node given a new reference, and an API in `TopicGraph` to update a node's reference.

**Comment by franklinsch** [comment]
> Was tutorials the last thing we needed to support here or do we still need support for articles?

I believe tutorials was the last thing we needed to make multi-language. Article support was added by #81.

**Comment by franklinsch** [comment]
@swift-ci please test

**Comment by franklinsch** [comment]
@swift-ci please test Linux platform

File: Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift (hunk 11 above)
Comment (ethan-kusters) [reply Sources/SwiftDocC/Model/Rendering/RenderNodeTranslator.swift]: Deal. 🤝😃

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift [hunk 13]
```diff
@@ -621,15 +624,30 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
...
+                    sourceLanguage: Self.defaultLanguage(in: sourceLanguages),
+                    availableSourceLanguages: sourceLanguages,
+                    name: .conceptual(title: technology.intro.title),
+                    markup: technology.originalMarkup,
+                    semantic: technology
+                )
+                documentationCache[technologyReference] = technologyNode
+                
+                // Update the reference in the topic graph with the technology's available languages.
+                topicGraph.updateReference(
```
Comment (ethan-kusters) [inline Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:641]: Nice! This looks great- much easier for me to follow.

**Review by ethan-kusters — APPROVED** [review]
